Develop a conversation simulator

jlaframboise opened the pull request:
# What
Added capabilities to simulate conversations with a chatbot. 

# Why
So sample conversations can be generated an evaluated.

# How
Using data in data.json which is one of the outputs from plan4dial we can intelligently select a response intent and an utterance. 

# Notes to reviewers
You can run this with `python local_main_simulated.py`

jlaframboise committed:
Added dockerfile

jlaframboise committed:
Merge branch 'create_bot_integrate_rollout' into dockerize

jlaframboise committed:
Implemented a simulator which instead of asking for user input to give to the chatbot, will sample responses based on the expected outcomes.

jlaframboise committed:
Merge branch 'dialogue-planning:main' into simulate

jlaframboise committed:
Revert local_main.py to its original state before the simulator to avoid changing others' workflows. Add a simulator specific script.

jlaframboise committed:
Name change for rasa model server was implemented in main, adding this change to my code.

jlaframboise committed:
Add one sample conversation.

jlaframboise committed:
Removed a deprecated function, and updated doc strings in a couple functions.

jlaframboise committed:
Small PR edits

jlaframboise committed:
Finished json output.

Made small ConversationLog classes to make custom output formats for conversations. Currently there is a json and text class for logs.

jlaframboise committed:
Removed dead code.

jlaframboise committed:
Merge branch 'dialogue-planning:main' into simulate

jlaframboise committed:
Filtered out fallback responses.

Using the recent addition to plan4dial which includes an intent type in data.json, the fallback intent is not randomly selected as a response so conversations progress as expected.

jlaframboise committed:
Update example conversations.

beckydvn commented on hovor/actions/local_dialogue_action_simulated.py:
```diff
@@ -0,0 +1,195 @@
+        #     intents = yaml.safe_load(fin)
+        # self.intents = intents['nlu']
+        # This is used in the process of generating a response
+        with open("/home/jacob/Dev/plan4dial/plan4dial/local_data/gold_standard_bot/output_files/data.json") as fin:
```
Can this be replaced with an output files parameter? (Crashes on local runs)

jlaframboise replied:
I agree it isn't great the have that hardcoded. I have spent some time looking at how to parameterize it and it is proving to be tricky, because the call stack
![image](https://user-images.githubusercontent.com/38571487/213943781-e428b2e7-4319-4f5b-b9a6-14d86a1b72e7.png)

Goes through a number of files that I don't think should be modified for simulation. 

I settled on a change in json_configuration_provider to include a copy of the data when the action is instantiated. I actually realize I don't need to even read from the file, that is actually already done bu hovor when specifying output files.

beckydvn commented on hovor/actions/local_dialogue_action_simulated.py:
```diff
@@ -0,0 +1,195 @@
+        return entity_dicts
+
+    def get_entity_names_from_string(self, example):
+        entity_dicts = self.get_entities_from_string(example)
```
If you are just trying to get the entity names (i.e. "budget" and "outing_type") you can get this from self.data["intents"]

beckydvn commented on hovor/actions/local_dialogue_action_simulated.py:
```diff
@@ -0,0 +1,195 @@
+        filters these examples to ones that have the same entities as the required entities,
+        and returns a random choice from the possible examples as the response.
+
+        This works well for share_cuisine or share_outing_type,
```
If this is still being used, it should be refactored so that "no required entities" is still a valid option. It looks like this function is accomplishing the same as `generated_response_by_data`, any reason why one isn't phased out for the other?

beckydvn replied:
Also, do you care to simulate any entities that are "maybe" extracted? An easy way you could do this is to randomly change, add, or delete a letter in one of the "valid" entity options, i.e. "piza" instead of "pizza." I have a plugged-in spell checker in Rasa such that if the extracted entity is valid after a spell check, it gets a "maybe."

beckydvn replied:
Another thing - this is probably me misunderstanding the wording but if you are picking a random outcome you shouldn't have to then pick the intent as there should be one intent linked to every outcome already.

Finally, if you do end up taking "maybe"s into account you will have to look at the entity assignments in each outcome to determine the status of each entity. This shouldn't break anything but also note that multiple outcomes can have the same intent but with different entity assignments (mostly the case for `slot_fill` actions).

jlaframboise replied:
Yes I agree I should remove the generate by required entities function -- I don't forsee a case where that would be a better approach.

jlaframboise replied:
Also yes, this function was made before I started using data.json so it was trying to work around not having self.data defined. I will remove this function.

jlaframboise replied:
The maybe's are an interesting idea. My first instinct is that I am primarily interested in when the chatbot does something weird upon normal inputs so I shouldn't simulate abnormal inputs, but there would definitely be value in looking at it. I will give it more thought, it might be out of the scope of my project.

beckydvn replied:
All good, if you think it is worth pursuing at some point, perhaps create an issue for it, but we can move past it for now :)

beckydvn commented on local_main_simulated.py:
```diff
@@ -0,0 +1,15 @@
+from local_run_utils import *
+
+
+def run_local_conversation(output_files_path):
```
Does this function execute any differently than `simulate_local_conversation`?

jlaframboise replied:
it calls `initialize_local_run_simulated` instead of `initialize_local_run`

beckydvn reviewed (commented):
Awesome work :)

jlaframboise committed:
Merge branch 'dialogue-planning:main' into simulate

jlaframboise committed:
Reworked data.json loading and removed deprecated functions.

Hardcoding the data.json path in LocalDialogueAction was bad practice, instead I find this data in the configuration attribute that hovor loaded and add this into the config for the action. So no extra file read or path input is needed.

Since the old method of generating a response from required entities is worse than the new one, I removed it.
Since a lot of functions supported that and are also not longer needed, they were removed too.

beckydvn commented on hovor/actions/local_dialogue_action_simulated.py:
```diff
@@ -0,0 +1,139 @@
+import random
+import re
+import yaml
```
add to requirements.txt

jlaframboise replied:
Actually I can delete the yaml and json import, no longer needed as I am reading from memory not file after most recent changes.

beckydvn commented on hovor/actions/local_dialogue_action_simulated.py:
```diff
@@ -0,0 +1,139 @@
+            else:
+                # must be fallback, there are no utterances for this intent
+                simulated_input = "I'm not sure."
+        elif isinstance(random_outcome_intent_name, dict):
```
If this is used to handle "intent dicts", I actually ended up phasing them out recently as they ended up being problematic for other reasons. So you are probably good to delete this whole block :)

jlaframboise replied:
Oh okay that makes sense! When I was implementing the last changes with you recent changes pulled in, I couldn't seem to see where the dict intents would occur but I just knew I had encountered them. I will remove this then

beckydvn reviewed (dismissed):
Awesome work -- Just a couple tiny changes left and you're good to merge! (Also have a look at creating the issue I mentioned above if you think it's worth it)

jlaframboise committed:
Simplify simulated action.

Since dictionary intents were removed, the generate response by data no longer needs to handle the dictionary intents.

Since no output files are being read from file in the diologue action anymore, the imports for json and yaml are no longer needed.
